src/category_theory/closed/ideal.lean

```diff
@@ -0,0 +1,278 @@
+/-
+Copyright (c) 2021 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+import category_theory.limits.preserves.shapes.binary_products
+import category_theory.limits.shapes.constructions.finite_products_of_binary_products
+import category_theory.monad.limits
+import category_theory.adjunction.fully_faithful
+import category_theory.adjunction.reflective
+import category_theory.closed.cartesian
+import category_theory.subterminal
+
+/-!
+# Exponential ideals
+
+An exponential ideal of a cartesian closed category `C` is a subcategory `D ⊆ C` such that for any
+`B : D` and `A : C`, the exponential `A ⟹ B` is in `D`: resembling ring theoretic ideals. We
+define the notion here for inclusion functors `i : D ⥤ C` rather than explicit subcategories to
+preserve the principle of equivalence.
+
+We additionally show that if `C` is cartesian closed and `i : D ⥤ C` is a reflective functor, the
+following are equivalent.
+* The left adjoint to `i` preserves binary (equivalently, finite) products.
+* `i` is an exponential ideal.
+-/
+universes v₁ v₂ u₁ u₂
+
+noncomputable theory
+
+namespace category_theory
+
+open limits category
+
+section ideal
+
+variables {C : Type u₁} {D : Type u₂} [category.{v₁} C] [category.{v₁} D] {i : D ⥤ C}
+
+variables (i) [has_finite_products C] [cartesian_closed C]
+
+/--
+The subcategory `D` of `C` expressed as an inclusion functor is an *exponential ideal* if
+`B ∈ D` implies `A ⟹ B ∈ D` for all `A`.
+-/
+class exponential_ideal : Prop :=
+(exp_closed : ∀ {B}, B ∈ i.ess_image → ∀ A, (A ⟹ B) ∈ i.ess_image)
+
+/--
+To show `i` is an exponential ideal it suffices to show that `A ⟹ iB` is "in" `D` for any `A` in
+`C` and `B` in `D`.
+-/
+lemma exponential_ideal.mk' (h : ∀ (B : D) (A : C), (A ⟹ i.obj B) ∈ i.ess_image) :
+  exponential_ideal i :=
+⟨λ B hB A,
+begin
+  rcases hB with ⟨B', ⟨iB'⟩⟩,
+  exact functor.ess_image.of_iso ((exp A).map_iso iB') (h B' A),
+end⟩
+
+/-- The subcategory of subterminal objects is an exponential ideal. -/
+instance : exponential_ideal (subterminal_inclusion C) :=
```

Comment:
I guess the other "canonical" example is the identity functor? Could you please add that?

Reply:
Added, thanks!